### PR TITLE
Add hidden "Test Room" to public library for TestCafe

### PR DESCRIPTION
**Pull request by 96LawDawg**
This PR allows rooms to be added as a TestCafe test by being added to the Public Library, but without being added to the list of Public Library games and Tutorials.

Instead of the test rooms being added through the public library list (library.json), they are added through the "Enter Link" button.

It also adds one test room file "Test_Room.vtt" to the public library that runs a lot of functions and changes a lot of properties.

Additional test rooms can be added by adding the vtt file to the public library folder, and adding a line that follows the following format.

hiddenLibraryButtons('NameOfFileExcludingTheDotvtt',      Variant#, 'Hash', [
  'IdOfButton1', 'IdOfButton2', 'IdOfButton3'
]);

This is an example with one button to click.

hiddenLibraryButtons('Test_Room',      0, 'fe86952d067fc69feb6dff7937f6b548', [
  'clickThis'
]);

This PR also updates public library tests to remove games if they remained after a previous run.  This seems to work, since it got rid of the errors showing up in Blue (presumably because a Test Room remained).  Hopefully it will prevent what appeared to be random failures previously.

**Problems:**
On my localhost server this is passing in TestCafe in both Edge and Chrome using the commands `testcafe edge tests/testcafe/tests.js` & `testcafe chrome tests/testcafe/tests.js`.  

image

I did reproduce the error using the command `npx testcafe chrome:headless tests/testcafe/tests.js` & with the command `testcafe chrome:headless tests/testcafe/tests.js`.

However it passes with the command `npx testcafe chrome tests/testcafe/tests.js'

image

Only fails when **chrome:headless** is run.

Possibly related information:
https://github.com/DevExpress/testcafe/issues/5482
https://github.com/DevExpress/testcafe/issues/3862
https://github.com/DevExpress/testcafe/issues/4197
https://testcafe.io/documentation/402963/guides/advanced-guides/speed-up-test-execution#run-tests-in-headless-browsers
https://testcafe.io/documentation/402828/guides/concepts/browsers#test-in-headless-mode

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3724/pr-test (or any other room on that server)

**Comment by rogerl50** [comment]
So what is the process here? Should @bjalder26 update the TestCafe tests on his local environment, push those changes up here, and then once the tests all pass, the test room as well as the updated tests get added to main?

**Comment by 96LawDawg** [comment]
> So what is the process here? Should @bjalder26 update the TestCafe tests on his local environment, push those changes up here, and then once the tests all pass, the test room as well as the updated tests get added to main?

I think we treat it just like any other public library file.  It gets updated as needed to keep up with changes to the code and pushed to the website.  Should be simple enough once we get it to work (which it isn't doing right now).

**Comment by bjalder26** [comment]
Attempting to reinstall TestCafe by closing and reopening.

File: tests/testcafe/tests.js [hunk 1]
```diff
@@ -50,6 +50,35 @@ function prepareClient() {
...
+}
+
+function hiddenLibraryButtons(game, variant, md5, buttons) {
+  hiddenTest(game, variant, md5, async t => {
+    for(const b of buttons) {
+      await t.click(`[id="${b}"]`).wait(4000);
+    }
+  });
+}
+
```
Comment (ArnoldSmith86) [inline tests/testcafe/tests.js]: This is mostly just copied code. That's fine for testing but you should try to merge it with the other functions when it is working.

File: tests/testcafe/tests.js [hunk 2]
```diff
@@ -50,6 +50,35 @@ function prepareClient() {
...
+    await setName(t);
+    await tests(t);
+    await compareState(t, md5);
+  });
+}
+
+function hiddenLibraryButtons(game, variant, md5, buttons) {
+  hiddenTest(game, variant, md5, async t => {
+    for(const b of buttons) {
+      await t.click(`[id="${b}"]`).wait(4000);
```
Comment (ArnoldSmith86) [inline tests/testcafe/tests.js]: `wait(4000)` should not be necessary.

File: tests/testcafe/tests.js [hunk 3]
```diff
@@ -50,6 +50,35 @@ function prepareClient() {
...
+    await removeGame(t);
+    await t.expect(Selector('#statesOverlay').visible).ok();
+  })(`Public library: ${game} (variant ${variant})`, async t => {
+    await ClientFunction(prepareClient)();
+    await t
+      .pressKey('esc')
+      .click('#statesButton')
+      .setNativeDialogHandler(() => testUrl, { dependencies: { testUrl }})
+      .click('button.prettyButton.link:nth-child(2)')
+      .pressKey('enter')
```
Comment (ArnoldSmith86) [inline tests/testcafe/tests.js]: The first line already adds the game. The second line adds it again which is the reason why one of those two games was still there when running TestCafé again.

You can see this easily by running `npm run testcafe-firefox -- --test-grep '.*Test_Room.*' --debug-mode`.

File: tests/testcafe/tests.js [hunk 4]
```diff
@@ -59,7 +88,12 @@ function publicLibraryTest(game, variant, md5, tests) {
     await t
       .pressKey('esc')
       .click('#statesButton')
-      .click(Selector('td.name').withExactText(game).prevSibling().child())
+    if(!await Selector('td.name').visible) {
+      await t.hover(`.roomState:nth-of-type(1)`)
+        .click(`.roomState:nth-of-type(1) .edit`)
+        .click('p > .remove');
+    }
+    await t.click(Selector('td.name').withExactText(game).prevSibling().child())
```
Comment (ArnoldSmith86) [inline tests/testcafe/tests.js]: This shouldn't be necessary. See comment above. The main job of TestCafé is automated testing inside GitHub. There it gets a clean environment every time it is run.

When working on the TestCafé tests, you might have to cleanup the room manually by going to `http://localhost:8272/testcafe-testing`.

File: tests/testcafe/tests.js [hunk 5]
```diff
@@ -332,3 +366,7 @@ publicLibraryButtons('Master Button',      0, 'eb19dffdb38641d5556e5fdb2c47c62b'
   'masterbutton', 'redbutton', 'orangebutton', 'yellowbutton', 'greenbutton', 'bluebutton', 'indigobutton',
   'violetbutton', 'fae4', 'vbx5'
 ]);
+
+hiddenLibraryButtons('Test_Room',     0, '44784f8e8957d15535b7bf04ea9a72c3', [
+  'clickThis', 'button1', 'button2', 'type', 'button3', 'button4'
```
Comment (ArnoldSmith86) [inline tests/testcafe/tests.js]: I don't see a widget with ID `type`. Either you change it in a routine somewhere or you just messed up when editing the button (it has `type: "type"`).

**Review by ArnoldSmith86 — CHANGES_REQUESTED** [review]
I changed some stuff but did not have permission to commit it (at least from GitPod). Here's my file: [tests.js.zip](https://github.com/ArnoldSmith86/virtualtabletop/files/7006365/tests.js.zip)

In general this is very nice. A few things I noticed:
- Keep in mind that TestCafé only compares the resulting room state after all your buttons were pressed. Visual changes to the widgets will not be tested (only the properties they come from).
- If two of your buttons change the same property, the first change will never be checked because it only compares the state at the end.
- You seem to have a timer running. If the timer has a different time at the end of the test on different machines, the hash comparison will fail because the resulting room state is not the same.

**Comment by ArnoldSmith86** [comment]
Oh, my commit was actually added. I didn't realize. GitPod told me it didn't have permission...

File: tests/testcafe/tests.js (hunk 2 above)
Comment (bjalder26) [reply tests/testcafe/tests.js]: It might not be, but I put it in there to give the timer time to tick away.  I'll double check though.

Comment (ArnoldSmith86) [reply tests/testcafe/tests.js]: I see. Look at my comment about the timer though. It's a good idea to check if the timer was running by checking `milliseconds > X` in your last button (not sure if it gets enough time to tick) but you'll have to set the `milliseconds` to something predictable after that test (and put the test result into a property so the state comparison can find it).

File: tests/testcafe/tests.js (hunk 5 above)
Comment (bjalder26) [reply tests/testcafe/tests.js]: image
It has a type and id of "type".  However, I'm waiting for an update to the library file.  I made a more simplified version that passes on my home server.

File: tests/testcafe/tests.js (hunk 1 above)
Comment (bjalder26) [reply tests/testcafe/tests.js]: Agreed.

File: tests/testcafe/tests.js (hunk 3 above)
Comment (bjalder26) [reply tests/testcafe/tests.js]: I thought the .click would just open up the 'Enter Link" dialog box, and that enter would be required after the text was entered.  But if it is working, then I was wrong.

File: tests/testcafe/tests.js (hunk 2 above)
Comment (bjalder26) [reply tests/testcafe/tests.js]: It stops at a specific time.  In the most recent iteration it only goes for a couple of seconds.

**Comment by bjalder26** [comment]
This alone tells me that chrome:headless is running an older copy of the room: "nullcall".  Closing and reopening in hopes it will update.

**Comment by bjalder26** [comment]
I'm not sure what to do about this.  What this shows me is that firefox:headless is running and "passing" on an old copy of the room, and with an old hash that matches that room.

chrome:headless however, is actually performing the test correctly on the updated copy of the room, thus "failing" with the old hash.

When the green one starts with "nullcall" that tells me that 1) it was before I replaced a SET/+ with a SET/=, and it was also before I fixed it so the timer functions would have a listing in my "record".  In other words, it is an old copy of the room.

I saw a similar problem on my home PC where I could even watch a test being performed, check the "record" manually, then see a different "record" show up in the error report (from an older copy of the room).  So I am not sure if the old copy of the room is still being used by firefox:headless, or if it just keeps sending the wrong/old hash.

image

Also, the "record" tells me that all the buttons are being clicked in chrome:headless before the timer has a chance to run completely.  However, it looks like the timer IS running completely.  So I am not sure if that is even an issue.  My fear would be that the timer would sometimes complete before one of the buttons, but I don't think that is highly likely.

**Comment by 96LawDawg** [comment]
Going to open another, clean version of this.

**Pull request closed without merging**